src/jit/optimizer.cpp

```diff
@@ -6203,6 +6203,12 @@ bool Compiler::optHoistLoopExprsForTree(
         {
             treeIsHoistable = false;
         }
+
+        // Avoid to hoist reference loading
+        if (tree->OperGet() == GT_CLS_VAR && tree->gtType == TYP_REF)
+        {
+            treeIsInvariant = false;
```

Comment:
I would add a comment 

// It is unsafe to make this as invariant here as we currently don't prevent it from being hoisted above the call to the class initializer.